## Predicting the Stock Market

We'll be using historical data on the price of the S&P500 Index to make predictions about future prices. Predicting whether an index will go up or down will help us forecast how the stock market as a whole will perform. Since stocks tend to correlate with how well the economy as a whole is performing, it can also help us make economic forecasts.
We'll be using this dataset to develop a predictive model. We'll train the model with data from 1950-2012, and try to make predictions from 2013-2015.

In [1]:
import pandas as pd
from datetime import datetime

df = pd.read_csv('sphist.csv')
df['Date'] = pd.to_datetime(df['Date'])

df.head()

Date         Open         High          Low        Close  \
0 2015-12-07  2090.419922  2090.419922  2066.780029  2077.070068   
1 2015-12-04  2051.239990  2093.840088  2051.239990  2091.689941   
2 2015-12-03  2080.709961  2085.000000  2042.349976  2049.620117   
3 2015-12-02  2101.709961  2104.270020  2077.110107  2079.510010   
4 2015-12-01  2082.929932  2103.370117  2082.929932  2102.629883   

         Volume    Adj Close  
0  4.043820e+09  2077.070068  
1  4.214910e+09  2091.689941  
2  4.306490e+09  2049.620117  
3  3.950640e+09  2079.510010  
4  3.712120e+09  2102.629883

In [2]:
df = df.sort_values(by=['Date'], ascending=True)
df.head()

Date   Open   High    Low  Close     Volume  Adj Close
16589 1950-01-03  16.66  16.66  16.66  16.66  1260000.0      16.66
16588 1950-01-04  16.85  16.85  16.85  16.85  1890000.0      16.85
16587 1950-01-05  16.93  16.93  16.93  16.93  2550000.0      16.93
16586 1950-01-06  16.98  16.98  16.98  16.98  2010000.0      16.98
16585 1950-01-09  17.08  17.08  17.08  17.08  2520000.0      17.08

### Creating Indicators

Datasets taken from the stock market need to be handled differently than datasets from other sectors when it comes time to make predictions. This means we have to be extra careful to not inject "future" knowledge into past rows when we do training and prediction. Injecting future knowledge will make our model look good when we're training and testing it, but will make it fail in the real world. The time series nature of the data means that can generate indicators to make our model more accurate.

Here are some indicators that are interesting to generate for each row:

- The average price from the past 5 days.
- The average price for the past 30 days.
- The average price for the past 365 days.
- The ratio between the average price for the past 5 days, and the average price for the past 365 days.
- The standard deviation of the price over the past 5 days.
- The standard deviation of the price over the past 365 days.
- The ratio between the standard deviation for the past 5 days, and the standard deviation for the past 365 days.
- The average volume over the past five days.
- The average volume over the past year.
- The ratio between the average volume for the past five days, and the average volume for the past year.
- The standard deviation of the average volume over the past five days.
- The standard deviation of the average volume over the past year.
- The ratio between the standard deviation of the average volume for the past five days, and the standard deviation of the average volume for the past year.
- The year component of the date.
- The ratio between the lowest price in the past year and the current price.
- The ratio between the highest price in the past year and the current price.
- The year component of the date.
- The month component of the date.
- The day of week.
- The day component of the date.
- The number of holidays in the prior month.


In [3]:
# Average price from the past 5 days (present day exluded)
df['avg_close_5'] = df.Close.rolling(5).mean().shift(1, axis=0)

# Average price from the past 30 days (present day exluded)
df['avg_close_30'] = df.Close.rolling(30).mean().shift(1, axis=0)

# Standard deviation of the price over the past 365 days (present day exluded)
df['std_close_365'] = df.Close.rolling(365).std().shift(1, axis=0)

# Average volume from the past 30 days (present day exluded)
df['avg_vol_5'] = df.Volume.rolling(5).mean().shift(1, axis=0)

df.head(50)

Date       Open       High        Low      Close     Volume  \
16589 1950-01-03  16.660000  16.660000  16.660000  16.660000  1260000.0   
16588 1950-01-04  16.850000  16.850000  16.850000  16.850000  1890000.0   
16587 1950-01-05  16.930000  16.930000  16.930000  16.930000  2550000.0   
16586 1950-01-06  16.980000  16.980000  16.980000  16.980000  2010000.0   
16585 1950-01-09  17.080000  17.080000  17.080000  17.080000  2520000.0   
16584 1950-01-10  17.030001  17.030001  17.030001  17.030001  2160000.0   
16583 1950-01-11  17.090000  17.090000  17.090000  17.090000  2630000.0   
16582 1950-01-12  16.760000  16.760000  16.760000  16.760000  2970000.0   
16581 1950-01-13  16.670000  16.670000  16.670000  16.670000  3330000.0   
16580 1950-01-16  16.719999  16.719999  16.719999  16.719999  1460000.0   
16579 1950-01-17  16.860001  16.860001  16.860001  16.860001  1790000.0   
16578 1950-01-18  16.850000  16.850000  16.850000  16.850000  1570000.0   
16577 1950-01-19  16.870001  16.870001  16.870001  16.870001  1170000.0   
16576 1950-01-20  16.900000  16.900000  16.900000  16.900000  1440000.0   
16575 1950-01-23  16.920000  16.920000  16.920000  16.920000  1340000.0   
16574 1950-01-24  16.860001  16.860001  16.860001  16.860001  1250000.0   
16573 1950-01-25  16.740000  16.740000  16.740000  16.740000  1700000.0   
16572 1950-01-26  16.730000  16.730000  16.730000  16.730000  1150000.0   
16571 1950-01-27  16.820000  16.820000  16.820000  16.820000  1250000.0   
16570 1950-01-30  17.020000  17.020000  17.020000  17.020000  1640000.0   
16569 1950-01-31  17.049999  17.049999  17.049999  17.049999  1690000.0   
16568 1950-02-01  17.049999  17.049999  17.049999  17.049999  1810000.0   
16567 1950-02-02  17.230000  17.230000  17.230000  17.230000  2040000.0   
16566 1950-02-03  17.290001  17.290001  17.290001  17.290001  2210000.0   
16565 1950-02-06  17.320000  17.320000  17.320000  17.320000  1490000.0   
16564 1950-02-07  17.230000  17.230000  17.230000  17.230000  1360000.0   
16563 1950-02-08  17.209999  17.209999  17.209999  17.209999  1470000.0   
16562 1950-02-09  17.280001  17.280001  17.280001  17.280001  1810000.0   
16561 1950-02-10  17.240000  17.240000  17.240000  17.240000  1790000.0   
16560 1950-02-14  17.059999  17.059999  17.059999  17.059999  2210000.0   
16559 1950-02-15  17.059999  17.059999  17.059999  17.059999  1730000.0   
16558 1950-02-16  16.990000  16.990000  16.990000  16.990000  1920000.0   
16557 1950-02-17  17.150000  17.150000  17.150000  17.150000  1940000.0   
16556 1950-02-20  17.200001  17.200001  17.200001  17.200001  1420000.0   
16555 1950-02-21  17.170000  17.170000  17.170000  17.170000  1260000.0   
16554 1950-02-23  17.209999  17.209999  17.209999  17.209999  1310000.0   
16553 1950-02-24  17.280001  17.280001  17.280001  17.280001  1710000.0   
16552 1950-02-27  17.280001  17.280001  17.280001  17.280001  1410000.0   
16551 1950-02-28  17.219999  17.219999  17.219999  17.219999  1310000.0   
16550 1950-03-01  17.240000  17.240000  17.240000  17.240000  1410000.0   
16549 1950-03-02  17.230000  17.230000  17.230000  17.230000  1340000.0   
16548 1950-03-03  17.290001  17.290001  17.290001  17.290001  1520000.0   
16547 1950-03-06  17.320000  17.320000  17.320000  17.320000  1470000.0   
16546 1950-03-07  17.200001  17.200001  17.200001  17.200001  1590000.0   
16545 1950-03-08  17.190001  17.190001  17.190001  17.190001  1360000.0   
16544 1950-03-09  17.070000  17.070000  17.070000  17.070000  1330000.0   
16543 1950-03-10  17.090000  17.090000  17.090000  17.090000  1260000.0   
16542 1950-03-13  17.120001  17.120001  17.120001  17.120001  1060000.0   
16541 1950-03-14  17.250000  17.250000  17.250000  17.250000  1140000.0   
16540 1950-03-15  17.450001  17.450001  17.450001  17.450001  1830000.0   

       Adj Close  avg_close_5  avg_close_30  std_close_365  avg_vol_5  
16589  16.660000          NaN           NaN            NaN        NaN  
16588  16.850000          NaN           Na

### Dropping Rows with `NaN` and Splitting Dataset into Train and Test

In [4]:
df = df[df["Date"] > datetime(year=1951, month=1, day=2)].dropna(axis=0)
df.isnull().sum()

Date             0
Open             0
High             0
Low              0
Close            0
Volume           0
Adj Close        0
avg_close_5      0
avg_close_30     0
std_close_365    0
avg_vol_5        0
dtype: int64

In [5]:
# The training set is composed of rows before 2013-01-01
train = df[df["Date"] < datetime(year=2013, month=1, day=1)]

# The test set is composed of rows after (including) 2013-01-01
test = df[df["Date"] >= datetime(year=2013, month=1, day=1)]

print(train.shape)
print(test.shape)

(15486, 11)
(739, 11)


### Training the Model

We will use the Linear Regression algorithm and assess its performance by using the Mean Absolute Error (MAE).

In [15]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error

model = LinearRegression()
features = ['avg_close_5', 'avg_vol_5', 'avg_close_30']
model.fit(train[features], train['Close'])
predictions = model.predict(test[features])
mae = mean_absolute_error(test['Close'], predictions)
print('MAE: ', mae)

MAE:  16.150885859834414


### Conclusion

Using as features `avg_close_5`, `avg_vol_5` and `avg_close_30`, the model displays a `MAE` of 16.2.